project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module

```diff
@@ -554,43 +554,106 @@ function c4m_user_og_form_user_login_alter(&$form, &$form_state, $form_id) {
+function c4m_user_og_unified_workflow_register_submit($form, &$form_state) {
   if (!isset($_SESSION['gid_join']) || !$gid = $_SESSION['gid_join']) {
     return;
   }
 
   $group = node_load($gid);
-  if ($form_state['build_info']['form_id'] == 'user_register_form') {
-    // Removing the original message of the user_register_form.
-    drupal_get_messages('status');
-    $message = t(
-      'Hey @name, thank you for submitting your details. Please activate your account by clicking the link sent to your e-mail to join the @group_title group.', array(
-      '@name' => c4m_user_name($form_state['user']),
-      '@group_title' => $group->title,
-    ));
-  }
-  else {
-    $membership_request_value = c4m_og_get_group_membership_request($group);
-    $messages = array(
-      'open' => t('You are now a member of this group'),
-      'moderated' => t('This is a moderated group. Your request is waiting for approval. Continue to explore group @group_title content', array('@group_title' => $group->title)),
-    );
-    $message = $messages[$membership_request_value];
+  // Removing the original message of the user_register_form.
+  drupal_get_messages('status');
+  $message = t(
+    'Hey @name, thank you for submitting your details. Please activate your account by clicking the link sent to your e-mail to join the @group_title group.', array(
+    '@name' => c4m_user_name($form_state['user']),
+    '@group_title' => $group->title,
+  ));
+
+  drupal_set_message($message);
+  $form_state['redirect'] = drupal_get_path_alias('node/' . $gid);
+
+  // Adding the uid to the session so we can use it later.
+  $_SESSION['gid_join_uid'] = $form_state['user']->uid;
+
+  // drupal_goto has a fallback to the 'destination' so we must remove it.
+  unset($_GET['destination']);
+}
+
+/**
+ * A submit handler for the unified workflow (login).
+ *
+ * Setting a message to the user and redirect back to the group.
+ */
+function c4m_user_og_unified_workflow_login_submit($form, &$form_state) {
+  if (!isset($_SESSION['gid_join']) || !$gid = $_SESSION['gid_join']) {
+    return;
   }
 
+  $group = node_load($gid);
+  $membership_request_value = c4m_og_get_group_membership_request($group);
+  $messages = array(
+    'open' => t('You are now a member of this group'),
+    'moderated' => t('This is a moderated group. Your request is waiting for approval. Continue to explore group @group_title content', array('@group_title' => $group->title)),
+  );
+  $message = $messages[$membership_request_value];
+
   drupal_set_message($message);
   $form_state['redirect'] = drupal_get_path_alias('node/' . $gid);
 
   // drupal_goto has a fallback to the 'destination' so we must remove it.
   unset($_GET['destination']);
 }
 
+/**
+ * Implements hook_mail_alter().
+ */
+function c4m_user_og_mail_alter(&$message) {
+  if (!isset($_SESSION['gid_join']) || !$gid = $_SESSION['gid_join']) {
+    return;
+  }
+
+  if (empty($_SESSION['gid_join_uid'])) {
+    return;
+  }
+
```

Comment:
Isn't it risky to not check for the type of email sent too?
Another problem that I see is that if the email is resent through People ("admin/people") overview page ("Resend activation email" action) the email is not altered.